File: trees/111. Minimum Depth of Binary Tree/minDepth.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function(root) {
    if (!root) return 0;

    let stack = [];
    let min_path = Infinity;

    // Stack contains the node and the its depth in the tree
    stack.push([root, 1]);

    while (stack.length > 0) {
        let [current, curr_depth] = stack.pop();

        // If the current node is a leaf and it's the new smallest path
        if (!current.right && !current.left && curr_depth <= min_path) {
            min_path = curr_depth;
        }

        if (current.right) stack.push([current.right, curr_depth + 1]);
        if (current.left) stack.push([current.left, curr_depth + 1]);
    }

    return min_path;
};

File: trees/111. Minimum Depth of Binary Tree/minDepth.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def minDepth(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        if not root:
            return 0

        stack = [(root, 1)]
        min_path = float('inf')

        while stack:
            node, depth = stack.pop()
            if not node.left and not node.right:
                min_path = min(min_path, depth)
            if node.right:
                stack.append((node.right, depth + 1))
            if node.left:
                stack.append((node.left, depth + 1))

        return min_path